File: src/main/java/no/fint/provider/springer/model/OkonomiKodeverkRepository.java
package no.fint.provider.springer.model;

import lombok.extern.slf4j.Slf4j;
import no.fint.event.model.Event;
import no.fint.event.model.ResponseStatus;
import no.fint.event.model.Status;
import no.novari.fint.model.okonomi.kodeverk.KodeverkActions;
import no.novari.fint.model.resource.FintLinks;
import no.novari.fint.model.resource.okonomi.kodeverk.MerverdiavgiftResource;
import no.novari.fint.model.resource.okonomi.kodeverk.VareResource;
import no.fint.provider.springer.storage.SpringerRepository;
import org.springframework.stereotype.Repository;

import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Slf4j
@Repository
public class OkonomiKodeverkRepository extends SpringerRepository {

    @Override
    public void accept(Event<FintLinks> response) {
        log.debug("OkonomiKodeverk-handler request accept for " + response.getAction());

        switch (KodeverkActions.valueOf(response.getAction())) {
            case GET_ALL_MERVERDIAVGIFT:
                log.debug("Start fetching merverdiavgift");
                query(MerverdiavgiftResource.class, response);
                log.debug("Added " + response.getData().stream().count() + " merverdiavgif");
                break;
            case GET_ALL_VARE:
                log.debug("Start fetching vare");
                query(VareResource.class, response);
                log.debug("Added " + response.getData().stream().count() + " varer");
                break;
            default:
                log.debug("OkonomiKodeverk-handler request default - REJECTED");
                response.setStatus(Status.ADAPTER_REJECTED);
                response.setResponseStatus(ResponseStatus.REJECTED);
                response.setStatusCode("INVALID_ACTION");
                response.setMessage("Invalid action");
        }
    }

    @Override
    public Set<String> actions() {
        return Stream
                .of(KodeverkActions.values())
                .map(Enum::name)
                .filter(s -> s.startsWith("GET_ALL_"))
                .collect(Collectors.toSet());
    }
}

File: src/main/kotlin/no/fint/provider/springer/model/OkonomiKodeverkRepository.kt
package no.fint.provider.springer.model

import jakarta.annotation.PostConstruct
import no.fint.event.model.Event
import no.fint.event.model.ResponseStatus
import no.fint.event.model.Status
import no.fint.provider.springer.storage.SpringerRepository
import no.fint.provider.springer.storage.Wrapper
import no.novari.fint.model.okonomi.kodeverk.KodeverkActions
import no.novari.fint.model.resource.FintLinks
import no.novari.fint.model.resource.okonomi.kodeverk.*
import org.springframework.data.mongodb.core.MongoTemplate
import org.springframework.stereotype.Repository
import java.util.*
import java.util.stream.Collectors
import java.util.stream.Stream

@Repository
class OkonomiKodeverkRepository(
    mongoTemplate: MongoTemplate,
    wrapper: Wrapper
) : SpringerRepository(mongoTemplate, wrapper) {

    private val actionsMap: EnumMap<KodeverkActions, Class<out FintLinks>> = EnumMap(KodeverkActions::class.java)

    @PostConstruct
    fun init() {
        actionsMap[KodeverkActions.GET_ALL_MERVERDIAVGIFT] = MerverdiavgiftResource::class.java
        actionsMap[KodeverkActions.GET_ALL_VARE] = VareResource::class.java
    }

    override fun accept(response: Event<FintLinks>) {
        val action = KodeverkActions.valueOf(response.action)
        val type = actionsMap[action]
        if (type != null) {
            query(type, response)
        } else {
            invalid(response)
        }
    }

    private fun invalid(response: Event<FintLinks>) {
        response.status = Status.ADAPTER_REJECTED
        response.responseStatus = ResponseStatus.REJECTED
        response.statusCode = "INVALID_ACTION"
        response.message = "Invalid action"
    }

    override fun actions(): Set<String> =
        Stream.of(*KodeverkActions.values())
            .map { it.name }
            .filter { it.startsWith("GET_ALL_") }
            .collect(Collectors.toSet())
}
